src/smoosh/src/smoosh_channel.erl

```diff
@@ -177,21 +198,108 @@ handle_info(check_window, State0) ->
     erlang:send_after(60 * 1000, self(), check_window),
     {noreply, FinalState};
 
+handle_info(persist_queue, State0) ->
+    #state{waiting=Queue} = State0,
+    write_state_to_file(State0),
+    smoosh_priority_queue:write_to_file(Queue),
+    Checkpoint = config:get_integer("smoosh", "state_checkpoint_interval_in_sec", ?DEFAULT_CHECKPOINT_INTERVAL_IN_SEC) * 1000,
+    erlang:send_after(Checkpoint, self(), persist_queue),
+    {noreply, State0};
+
 handle_info(pause, State0) ->
-    {ok, State} = code_change(nil, State0, nil),
+    State = maybe_open_queue(State0),
     {noreply, State#state{paused=true}};
 handle_info(unpause, State0) ->
-    {ok, State} = code_change(nil, State0, nil),
+    State = maybe_open_queue(State0),
     {noreply, maybe_start_compaction(State#state{paused=false})}.
 
-terminate(_Reason, _State) ->
+terminate(_Reason, #state{name=Name, waiting = Q}) ->
+    file:delete(active_file_name(Name)),
+    file:delete(starting_file_name(Name)),
+    if Q =/= nil ->
+        smoosh_priority_queue:close(Q);
+    true ->
+        nil
+    end,
     ok.
 
-code_change(_OldVsn, #state{}=State, _Extra) ->
-    {ok, State}.
+maybe_recover_state(#state{name=Name} = State) ->
+    Active = recover(active_file_name(Name)),
+    Starting = recover(starting_file_name(Name)),
+    DatabaseDir = config:get("couchdb", "database_dir"),
+    ViewDir = config:get("couchdb", "view_index_dir"),
+    DatabaseCompactFiles = filelib:wildcard("*.compact", DatabaseDir),
+    ViewCompactFiles = filelib:wildcard("*.compact", ViewDir),
```

Comment:
IRC wildcard is very inefficient in erlang. We should try to avoid it or measure the impact (memory and time) when there are 100K of files in the directory. [fold_files](https://www.erlang.org/doc/man/filelib.html#fold_files-5) is better alternative. Since it doesn't require storing all filenames in memory.